### ISSUE TITLE
Create abstract NotificationManager class

### ISSUE THREAD

**Issue by mattkrick** [ParabolInc/parabol#5748]
Slack & Mattermost have their own notification systems.
Adding MS Teams would make 3.
Time to abstract & refactor!
We'll want to do this before we start work on the MS Teams integration.

@jordanh to add thoughts & modify the AC below because he touched it last 😝  

See related comments in https://github.com/ParabolInc/parabol/pull/5550

~~AC~~ Implementation suggestions:
- [ ] Slack & Mattermost extend an abstract `NotificationManager` class (name TBD)
- [ ] There is a parent `SuperNotificationManager` (please pick a better name 🙂 ) that also implements NotificationManager + has a bunch of NotificationManagers in it & it calls them all (e.g. SuperNotificationManager.startMeeting() calls Slack.startMeeting() and Mattermost.startMeeting())
- [ ] internally, it'd probably be nice for public methods like `startMeeting` to call internal helper methods like `getAuth()` and `getNotificationPreferences()` and then only notify the service if the caller has authentication & they want notifications sent for that particular notification type. If `auth` and `getNotificationPreferences` got cached inside the class even better, but it's up to the implementor to decide what's best.
 
~~Estimate: 8 hours~~

**Comment by jordanh** [comment]
Rad. Thank you for making this issue.

Rather than jump to implementation, let's ask a team member to propose a design first. I want to see those abstract methods and return types!

I've created and added a label called `architecture` for issues like this.

AC for this issue will be:
  - [ ] Class architecture with methods, implementation considerations (e.g. caching), and return types written
  - [ ] Reviewed with at least 1 Reviewer
  - [ ] Reviewed with at least 1 Maintainer

EE: 13 hours

I'm hoping this will save some time in implementation and reviewing when it comes to writing code!

**Comment by tiffanyhan** [comment]
interested

**Comment by tiffanyhan** [comment]
**WIP output from 12/15 sync jam session with Jordan:**

```
# could also be a string literal union
enum NotifyResponseEnum {
  SENT_SUCCESSFULLY = 'SENT_SUCCESSFULLY',
  NO_OP = 'NO_OP'
}
```
```
abstract class NotificationManager

  public startMeeting(meetingId, teamId, dataloader): Promise<NotifyResponseEnum | Error>
  public endMeeting(meetingId, teamId, dataloader): Promise<NotifyResponseEnum | Error>
  public startTimeLimit(meetingId, teamId, dataloader): Promise<NotifyResponseEnum | Error>
  public endTimeLimit(meetingId, teamId, dataloader): Promise<NotifyResponseEnum | Error>
  # TODO: should have a generic method to test if connection is valid

class SlackManager extends NotificationManager
class MattermostManager extends NotificationManager
class TeamsManager extends NotificationManager
```
```
class IntegratedServicesNotifier

  managers: NotificationManager[]

  public startMeeting(meetingId, teamId, dataloader): Promise<void>
  public endMeeting(meetingId, teamId, dataloader): Promise<void>
  public startTimeLimit(meetingId, teamId, dataloader): Promise<void>
  public endTimeLimit(meetingId, teamId, dataloader): Promise<void>
```

The reason for returning `Promise<Enum | Error>` from `NotificationManager` hook methods is bc sometimes it's helpful for the caller to know whether a service was sent, whether send failed, or whether it didn't even try to send (e.g. see [code here](https://github.com/ParabolInc/parabol/blob/de41393240bec3edd3999f22c450d70510f69de8/packages/server/graphql/intranetSchema/mutations/runScheduledJobs.ts#L56-L79)).

`NotificationManager` can also have private helper methods to get auth and communicate with the api. Since they aren't public, these methods aren't included in the spec, but there are [examples of such methods existing in the codebase](https://github.com/ParabolInc/parabol/blob/de41393240bec3edd3999f22c450d70510f69de8/packages/server/graphql/mutations/helpers/notifications/notifyMattermost.ts#L23-L56) already.

>  internally, it'd probably be nice for public methods like `startMeeting` to call internal helper methods like `getAuth()` and `getNotificationPreferences()` and then only notify the service if the caller has authentication & they want notifications sent for that particular notification type. If `auth` and `getNotificationPreferences` got cached inside the class even better, but it's up to the implementor to decide what's best.

per the above, I think it's probably better to keep the child as caller of its own `getAuth` and `getNotificationPreferences` methods for now. that way, `IntegratedServicesNotifier` can just call `NotificationManager`'s public methods, and be dumb to what is going on in those calls. I think it's better to KISS and keep control flow logic out of `IntegratedServicesNotifier`. We might refactor this later, but it feels like we don't know enough yet to abstract this out now.

Regarding cacheing inside the class, I prefer to keep servers as stateless as possible so we don't make horizontally scaling harder for us later. Integration configuration is unlikely to change often, which may make it a good candidate for cacheing. However, notification integrations are not particularly "hot". We might send a notification once every hour for a team. I think we can also table this for now for this first iteration.

_Other out of scope considerations_

- Can this abstraction also help us with in-app notifications, email notifications?
- Authentication: how can we refactor authentication so that it's generic and works for any possible self-hosted/oauth/webhook integration? This spec hasn't addressed this, as it is agnostic to how implementation occurs. However, since [parallel work](https://github.com/ParabolInc/parabol/pull/5594) is occurring for this, implementation of this issue should follow and elaborate upon the newly proposed pattern.

Todos:
- [ ] Review https://github.com/ParabolInc/parabol/pull/5594
- [ ] Watch loom https://www.loom.com/share/22037d51a1404fa5aab5d6ba10c98778
- [ ] look into files `packages/server/integrations/GitLabServerManager.ts` and `packages/server/integrations/IntegrationServerManager.ts` as entry points per Jordan's suggestion
- [ ] See if authentication can be implemented functional mix-in style or OOP

**Comment by tiffanyhan** [comment]
@jordanh anything missing from the spec below? since bartosz has been thinking about auth in https://github.com/ParabolInc/parabol/pull/5805, i think it's best we keep the discussion about that there, and this issue can be updated once the approach settles there.

```
# could also be a string literal union
enum NotifyResponseEnum {
  SENT_SUCCESSFULLY = 'SENT_SUCCESSFULLY',
  NO_OP = 'NO_OP'
}
```
```
interface NotificationIntegrationServerManager

  public startMeeting(meetingId, teamId, dataloader): Promise<NotifyResponseEnum | Error>
  public endMeeting(meetingId, teamId, dataloader): Promise<NotifyResponseEnum | Error>
  public startTimeLimit(meetingId, teamId, dataloader): Promise<NotifyResponseEnum | Error>
  public endTimeLimit(meetingId, teamId, dataloader): Promise<NotifyResponseEnum | Error>

class SlackManager implements NotificationIntegrationServerManager
class MattermostManager implements NotificationIntegrationServerManager
class TeamsManager implements NotificationIntegrationServerManager
```
```
class IntegratedServicesNotifier

  managers: NotificationIntegrationServerManager[]

  public startMeeting(meetingId, teamId, dataloader): Promise<void>
  public endMeeting(meetingId, teamId, dataloader): Promise<void>
  public startTimeLimit(meetingId, teamId, dataloader): Promise<void>
  public endTimeLimit(meetingId, teamId, dataloader): Promise<void>
```

The reason for returning `Promise<Enum | Error>` from `NotificationManager` hook methods is bc sometimes it's helpful for the caller to know whether a service was sent, whether send failed, or whether it didn't even try to send (e.g. see [code here](https://github.com/ParabolInc/parabol/blob/de41393240bec3edd3999f22c450d70510f69de8/packages/server/graphql/intranetSchema/mutations/runScheduledJobs.ts#L56-L79)).

each `NotificationManager` will also have helper methods to get auth and communicate with the api. these methods aren't included in the spec, and their implementations will depend on how the [add gitlab auth PR](https://github.com/ParabolInc/parabol/pull/5805) settles.
